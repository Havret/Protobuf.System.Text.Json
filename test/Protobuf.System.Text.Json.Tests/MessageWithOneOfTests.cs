using System.Text.Json.Protobuf.Tests.Utils;
using Shouldly;
using SmartAnalyzers.ApprovalTestsExtensions;
using Xunit;

namespace System.Text.Json.Protobuf.Tests;

public class MessageWithOneOfTests
{
    [Fact]
    public void Should_serialize_message_with_one_of_when_value_is_set()
    {
        // Arrange
        var msg = new MessageWithOneOf
        {
            Int32PropertyValue = 5
        };
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }

    [Fact]
    public void Should_serialize_message_with_one_of_when_value_is_not_set()
    {
        // Arrange
        var msg = new MessageWithOneOf();

        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }
    
    [Fact]
    public void Should_serialize_and_deserialize_message_with_one_of_when_value_is_set()
    {
        // Arrange
        var msg = new MessageWithOneOf
        {
            Int64PropertyValue = 5
        };
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithOneOf>(serialized, jsonSerializerOptions);

        // Assert
        deserialized.ShouldNotBeNull();
        deserialized.Int32PropertyValue.ShouldBe(msg.Int32PropertyValue);
        deserialized.Int64PropertyValue.ShouldBe(msg.Int64PropertyValue);
        deserialized.PropertyCase.ShouldBe(msg.PropertyCase);
    }
    
    [Fact]
    public void Should_serialize_and_deserialize_message_with_one_of_when_value_is_not_set()
    {
        // Arrange
        var msg = new MessageWithOneOf();
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithOneOf>(serialized, jsonSerializerOptions);

        // Assert
        deserialized.ShouldNotBeNull();
        deserialized.Int32PropertyValue.ShouldBe(msg.Int32PropertyValue);
        deserialized.Int64PropertyValue.ShouldBe(msg.Int64PropertyValue);
        deserialized.PropertyCase.ShouldBe(msg.PropertyCase);
    }
    
    [Fact]
    public void Should_serialize_message_with_optional_property_when_value_is_set()
    {
        // Arrange
        var msg = new MessageWithOptionalProperty
        {
            Int32Property = 5
        };
        
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }
    
    [Fact]
    public void Should_serialize_message_with_optional_property_when_value_is_not_set()
    {
        // Arrange
        var msg = new MessageWithOptionalProperty();
        
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }
    
    [Fact]
    public void Should_serialize_and_deserialize_message_with_optional_property_when_value_is_set()
    {
        // Arrange
        var msg = new MessageWithOptionalProperty()
        {
            Int32Property = 5
        };
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithOptionalProperty>(serialized, jsonSerializerOptions);

        // Assert
        deserialized.ShouldNotBeNull();
        deserialized.Int32Property.ShouldBe(msg.Int32Property);
        deserialized.HasInt32Property.ShouldBe(msg.HasInt32Property);
    }
    
    [Fact]
    public void Should_serialize_and_deserialize_message_with_optional_property_when_value_is_not_set()
    {
        // Arrange
        var msg = new MessageWithOptionalProperty();
        var jsonSerializerOptions = TestHelper.CreateJsonSerializerOptions();

        // Act
        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithOptionalProperty>(serialized, jsonSerializerOptions);

        // Assert
        deserialized.ShouldNotBeNull();
        deserialized.Int32Property.ShouldBe(msg.Int32Property);
        deserialized.HasInt32Property.ShouldBe(msg.HasInt32Property);
    }
}