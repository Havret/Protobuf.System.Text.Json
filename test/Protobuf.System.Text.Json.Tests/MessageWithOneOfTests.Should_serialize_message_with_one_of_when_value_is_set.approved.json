{
  "int32PropertyValue": 5
}